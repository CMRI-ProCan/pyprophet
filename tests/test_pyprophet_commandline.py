import os
import subprocess
import shutil
import sys

import pandas as pd

import pytest


d = pd.options.display
d.width = 220
d.precision = 6

DATA_FOLDER = os.path.join(os.path.dirname(os.path.abspath(__file__)), "data")


def _record(stdout, regtest):
    for line in stdout.split("\n"):
        if line.startswith("INFO"):
            # remove data which is not constant over runs as time or process id:
            line = line.split(":", 4)[-1]
        if "TIME" in line:
            continue
        print >> regtest, line


def _expected_output_files():
    return ["test_data_ms2_scored.tsv",
            "test_data_ms2_report.pdf",
            "test_data_ms2_weights.csv",
            "test_data_ms2_summary_stat.csv",
            "test_data_ms2_full_stat.csv",
            ]


def _remove_output_files(tmpdir, names=None):
    if names is None:
        names = _expected_output_files()
    for name in names:
        full_path = os.path.join(tmpdir.strpath, name)
        os.remove(full_path)


def _dump_output_files(tmpdir, regtest, names=None):
    try:
        tmpdir = tmpdir.strpath
    except AttributeError:
        pass
    if names is None:
        names = _expected_output_files()
    for name in names:
        full_path = os.path.join(tmpdir, name)
        if not name.endswith(".pdf"):
            _dump(full_path, regtest)


def _dump(full_path, regtest):
    head = open(full_path).readline()
    if "\t" in head:
        df = pd.read_csv(full_path, sep="\t", header=None)
    else:
        df = pd.read_csv(full_path, sep=" ", header=None)

    # lines = open(full_path, "r").readlines()
    # f = os.path.basename(full_path)
    if len(df) > 20:
        print >> regtest, df[:10]
        print >> regtest, "..."
        print >> regtest, df[-10:]
    else:
        print >> regtest, df


def _run_cmdline(cmdline):
    stdout = cmdline + "\n"
    try:
        stdout += subprocess.check_output(cmdline, shell=True,
                                          stderr=subprocess.STDOUT)
    except subprocess.CalledProcessError, e:
        print >> sys.stderr, e.output
        raise
    return stdout


def _run_pyprophet_to_learn_model(regtest, temp_folder, dump_result_files=False, parametric=False, pfdr=False, pi0_lambda=False):
    os.chdir(temp_folder)
    data_path = os.path.join(DATA_FOLDER, "test_data.txt")
    shutil.copy(data_path, temp_folder)
    cmdline = "pyprophet score --in=test_data.txt --test"
    if parametric:
        cmdline += " --parametric"
    if pfdr:
        cmdline += " --pfdr"
    if pi0_lambda is not False:
        cmdline += " --pi0_lambda=" + pi0_lambda
    stdout = _run_cmdline(cmdline)

    for f in _expected_output_files():
        full_path = os.path.join(temp_folder, f)
        assert os.path.exists(full_path)

    if not dump_result_files:
        return stdout

    _dump_output_files(temp_folder, regtest)

    return stdout


def test_tsv_0(tmpdir, regtest):
    stdout = _run_pyprophet_to_learn_model(regtest, tmpdir.strpath, True)
    _record(stdout, regtest)

def test_tsv_1(tmpdir, regtest):
    stdout = _run_pyprophet_to_learn_model(regtest, tmpdir.strpath, True, True)
    _record(stdout, regtest)

def test_tsv_2(tmpdir, regtest):
    stdout = _run_pyprophet_to_learn_model(regtest, tmpdir.strpath, True, False, True)
    _record(stdout, regtest)

def test_tsv_3(tmpdir, regtest):
    stdout = _run_pyprophet_to_learn_model(regtest, tmpdir.strpath, True, False, False, "0.3 0.55 0.05")
    _record(stdout, regtest)

def test_tsv_apply_weights(tmpdir, regtest):

    stdout = _run_pyprophet_to_learn_model(regtest, tmpdir.strpath, True)

    _record(stdout, regtest)

    full_path = os.path.join(tmpdir.strpath, "test_data_ms2_scored.tsv")
    _dump(full_path, regtest)

    # collect q-value stats
    m_score_stat, = [l for l in stdout.split("\n") if "mean q_value" in l]

    # split away log time etc:
    __, __, interesting_m_score_output = m_score_stat.partition("mean q_value")

    # collect s value stats
    s_value_stat, = [l for l in stdout.split("\n") if "mean s_value" in l]

    # split away log time etc:
    __, __, interesting_s_value_output = s_value_stat.partition("mean s_value")

    stdout = _run_cmdline("pyprophet score --in=test_data.txt --apply_weights=test_data_ms2_weights.csv "
                          "--test")

    _record(stdout, regtest)

    # collect m score stats
    m_score_stat, = [l for l in stdout.split("\n") if "mean q_value" in l]

    # split away log time etc:
    __, __, interesting_m_score_output2 = m_score_stat.partition("mean q_value")

    # collect s value stats
    s_value_stat, = [l for l in stdout.split("\n") if "mean s_value" in l]

    # split away log time etc:
    __, __, interesting_s_value_output2 = s_value_stat.partition("mean s_value")

    # check if output is consistent over runs:
    assert interesting_m_score_output == interesting_m_score_output2
    assert interesting_s_value_output == interesting_s_value_output2


def test_not_unique_tg_id_blocks(tmpdir):

    os.chdir(tmpdir.strpath)
    data_path = os.path.join(DATA_FOLDER, "test_invalid_data.txt")
    shutil.copy(data_path, tmpdir.strpath)
    cmdline = "pyprophet score --in=test_invalid_data.txt --test"

    with pytest.raises(subprocess.CalledProcessError) as exc_info:
        subprocess.check_output(cmdline, shell=True, stderr=subprocess.STDOUT)

    e = exc_info.value
    assert "Error: group_id values do not form unique blocks in input file(s)." in e.output
