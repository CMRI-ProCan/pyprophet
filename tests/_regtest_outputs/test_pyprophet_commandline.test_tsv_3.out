         0       1      2                               3                  4                            5                    6                              7                          8                 9   \
0  group_id  run_id  decoy  main_var_xx_swath_prelim_score  var_bseries_score  var_elution_model_fit_score  var_intensity_score  var_isotope_correlation_score  var_isotope_overlap_score  var_library_corr
1  459_run0       0      0                   5.27893963303                  5               0.993098149697       0.930480707464                 0.998724247981                        0.0     0.91240602174
2  459_run0       0      0                  -0.13974375122                  0               0.953924566507      0.0259317385401                 0.632511052964                        0.0   -0.285526965432
3  459_run0       0      0                  -0.77504285962                  0               0.666666249434     0.00044235703213                 0.657922811238                        0.0    0.376474234306
4  459_run0       0      0                  -1.05454511848                  0               0.847340305646      0.0009039469787                  0.79843565088             0.255319140851     0.51872168654
5  459_run0       0      0                    -1.246166932                  0                         -1.0     0.00042312411769                 0.563846670955            0.0454545468092    0.445906014642
6  459_run0       0      0                  -1.27951627249                  1               0.931647390127     0.00561216443373               -0.0419896690886            0.0997258424759  -0.0976738454584
7  459_run0       0      0                  -1.81518747764                  0               0.933585554361     0.00048082286101                 0.104774965663             0.159999996424   -0.630422997778
8  460_run0       0      0                   4.90480697454                  0               0.991616976261       0.929647682861                 0.998396899384                        0.0    0.757169232785
9  460_run0       0      0                   1.00012878546                  0               0.836714565754     0.00222462693417                 0.762015540501             0.380021706223    0.956728181447

        ...                          15                            16               17                        18                 19                20                 21                 22                 23  \
0       ...         var_xcorr_coelution  var_xcorr_coelution_weighted  var_xcorr_shape  var_xcorr_shape_weighted  var_yseries_score           d_score            p_value            q_value                pep
1       ...                         0.0                           0.0   0.960350100659            0.970979341002                  6      5.3194475174  5.20413026006e-08  7.11604145184e-09  3.72408611348e-08
2       ...               1.39668108475                0.859180469403   0.789175310515            0.716396888252                  0    -1.37402319908     0.908545913103    0.0632585334592                1.0
3       ...               2.38678570295                 1.97001683336   0.119047619048           0.0149915833201                  0    -5.57031345367     0.981405377553    0.0681548887411                1.0
4       ...               4.39320066984                 2.78930134498   0.329194873727            0.330811834697                  0    -3.48591995239     0.981405377553    0.0681548887411                1.0
5       ...                         1.0                           1.0              0.0                       0.0                  0    -1.27357709408     0.908545913103    0.0632585334592                1.0
6       ...               2.20848900055                 1.49086544255   0.711876219914            0.685971015368                  1    -3.50903868675     0.981405377553    0.0681548887411                1.0
7       ...               1.39696341001                0.799699939913   0.761807409832            0.743328859302                  0    -3.73713040352     0.981405377553    0.0681548887411                1.0
8       ...               2.17321312465                 1.00181054984   0.895313998806            0.867887743561                  5     4.81012630463   7.5417338663e-07  8.17295423233e-08  9.32678866022e-07
9       ...               3.54384793756                 1.31714642152   0.666435350514            0.636692707326                  0  -0.0955285355449     0.556997379025    0.0398129951979                1.0

                24
0  peak_group_rank
1                1
2                3
3                7
4                4
5                2
6                5
7                6
8                1
9                2

[10 rows x 25 columns]
...
                  0  1  2                3  4               5                 6                7                8                 9  ...             15              16              17              18 19  \
9156  DECOY_592_run0  0  1  -0.909754229584  0  0.823102881511   0.0126013214359  -0.147306602511   0.229729726911    0.778375379467 ...  3.94538540661   2.00615080523  0.604014629053  0.541176149999  2
9157  DECOY_592_run0  0  1  -0.956865429723  0            -1.0  0.00238403378516   0.266941597725              0.0  -0.0969535542793 ...            1.0             1.0             0.0             0.0  0
9158  DECOY_592_run0  0  1   -1.52418394133  0            -1.0  0.00017028812751  -0.583329727443              0.0  -0.0585549720413 ...            1.0             1.0             0.0             0.0  0
9159  DECOY_592_run0  0  1   -1.68998208875  0  0.895950257778  0.00854846400109   0.546651481853   0.601593613625   -0.189017878466 ...  2.06934963933   1.14626069677  0.785186892817  0.764795972809  0
9160  DECOY_592_run0  0  1     -1.692134624  0  0.958494941394  0.00630066071793  -0.328751834466   0.189189195633   -0.482215863506 ...  1.28809100151  0.791833291782  0.803144337956  0.768010655265  0
9161  DECOY_592_run0  0  1   -1.84945466301  0  0.633835911751  0.00561950820789   0.371593955459              0.0   -0.371771330281 ...  2.61135263418   2.02039942856   0.55597268955  0.330880632506  0
9162  DECOY_592_run0  0  1   -2.10957129817  0  0.906453510125  0.00527893195286    0.22376248428  0.0645161271095   -0.255514614107 ...  1.07777568371  0.598394985285  0.893420898864  0.890335674837  0
9163  DECOY_592_run0  0  1   -2.31560973872  0  0.905198852221   0.0028948981677   0.137664821691              0.0    0.876978922737 ...  3.43237660856   2.17985387492  0.539929223558   0.44766613662  0
9164  DECOY_592_run0  0  1   -2.53264045217  0  0.980351785819  0.00664123697296  0.0745622896143   0.205128207803    0.216855907728 ...  1.87132796282  0.908604380766  0.736385791197  0.708796452491  0
9165  DECOY_592_run0  0  1   -3.44062881815  0            -1.0  0.00102172876507  0.0339463918184   0.333333343267   0.0837371951856 ...            1.0             1.0             0.0             0.0  0

                   20              21               22   23  24
9156   -3.33646297455  0.981405377553  0.0681548887411  1.0  12
9157  -0.698309361935  0.774472434419  0.0544881605069  1.0   2
9158   -5.26060676575  0.981405377553  0.0681548887411  1.0  16
9159    -3.0108602047  0.981405377553  0.0681548887411  1.0  10
9160   -2.89471268654  0.981405377553  0.0681548887411  1.0   9
9161   -2.54190897942  0.981405377553  0.0681548887411  1.0   8
9162   -1.98239552975  0.981405377553  0.0681548887411  1.0   7
9163   -4.49693441391  0.981405377553  0.0681548887411  1.0  15
9164   -4.08524942398  0.981405377553  0.0681548887411  1.0  14
9165   -3.87155127525  0.981405377553  0.0681548887411  1.0  13

[10 rows x 25 columns]
                                                    0
0                                  score,weight,level
1   main_var_xx_swath_prelim_score,0.204607992554,ms2
2               var_bseries_score,-0.134398643255,ms2
3      var_elution_model_fit_score,-3.22538451141,ms2
4              var_intensity_score,0.929527276479,ms2
5     var_isotope_correlation_score,2.23882656776,ms2
6        var_isotope_overlap_score,-1.42184269353,ms2
7                var_library_corr,-0.591680798597,ms2
8                 var_library_rmsd,-12.5974249821,ms2
9               var_log_sn_score,-0.0293243248277,ms2
10              var_massdev_score,0.0112751823587,ms2
11     var_massdev_score_weighted,-0.026236520302,ms2
12               var_norm_rt_score,-11.7467204818,ms2
13             var_xcorr_coelution,0.183874853137,ms2
14   var_xcorr_coelution_weighted,-0.169770115711,ms2
15                  var_xcorr_shape,4.11084532588,ms2
16         var_xcorr_shape_weighted,2.88856669143,ms2
17               var_yseries_score,0.180666833457,ms2
                                                   0
0  qvalue,pvalue,svalue,pep,fdr,fnr,fpr,tp,tn,fp,...
1  0.0,3.68371999571e-13,0.0111072755578,7.285255...
2  0.00999999977648,0.161647022292,1.0,0.85329290...
3  0.019999999553,0.245013720483,1.0,1.0,0.017845...
4  0.0500000007451,0.698056428527,1.0,1.0,0.04924...
5  0.10000000149,0.981405377553,1.0,1.0,0.0681548...
6                           0.20000000298,,,,,,,,,,,
7                          0.300000011921,,,,,,,,,,,
8                           0.40000000596,,,,,,,,,,,
9                                     0.5,,,,,,,,,,,
                                                   0
0  cutoff,fdr,fn,fnr,fp,fpr,pvalue,qvalue,svalue,...
1  -2.54864907265,0.0681548887411,-0.499743218593...
2  -2.34406280518,0.0681548887411,-0.499743218593...
3  -2.1394765377,0.0681548887411,-0.499743218593,...
4  -1.93489038944,0.0681548887411,-0.499743218593...
5  -1.73030424118,0.0681548887411,-0.499743218593...
6  -1.52571809292,0.0632585334592,-1.45789124616,...
7  -1.32113182545,0.0632585334592,-1.45789124616,...
8  -1.11654567719,0.0607541303679,-1.48534496977,...
9  -0.911959528923,0.056588343462,-2.14576127198,...
...
                                                    0
42  5.83938455582,6.32782087473e-10,249.124314909,...
43  6.04397058487,2.31232321868e-10,274.124314858,...
44  6.24855709076,1.10266882598e-10,310.124314844,...
45  6.45314311981,5.14032678641e-11,331.12431484,0...
46  6.65772914886,2.12154842441e-11,342.124314839,...
47  6.86231565475,8.79445343712e-12,350.124314839,...
48  7.06690168381,4.84628868582e-12,355.124314839,...
49  7.27148771286,7.20886187965e-12,359.124314839,...
50  7.47607374191,7.20886187965e-12,359.124314839,...
51  7.6806602478,7.20886187965e-12,359.124314839,0...
pyprophet score --in=test_data.txt --test --pi0_lambda=0.3 0.55 0.05
 config settings:
     final_statistics.emp_p: False
     final_statistics.lambda: [ 0.3   0.35  0.4   0.45  0.5   0.55]
     final_statistics.lfdr_adj: 1.5
     final_statistics.lfdr_eps: 1e-08
     final_statistics.lfdr_monotone: True
     final_statistics.lfdr_transf: probit
     final_statistics.lfdr_trunc: True
     final_statistics.pfdr: False
     final_statistics.pi0_method: smoother
     final_statistics.pi0_smooth_df: 3
     final_statistics.pi0_smooth_log_pi0: False
     group_id: group_id
     is_test: True
     num_processes: 1
     semi_supervised_learner.initial_fdr: 0.15
     semi_supervised_learner.initial_lambda: [ 0.3   0.35  0.4   0.45  0.5   0.55]
     semi_supervised_learner.iteration_fdr: 0.02
     semi_supervised_learner.iteration_lambda: [ 0.3   0.35  0.4   0.45  0.5   0.55]
     semi_supervised_learner.num_iter: 10
     tric_chromprob: False
     xeval.fraction: 0.5
     xeval.num_iter: 10
 learn and apply classifier from input data
 Data set contains 387 decoy and 387 target groups.
 summary input file:
    9165 lines
    774 transition groups
    17 scores including main score
 learn and apply scorer
 start 10 cross evals using 1 processes
 start learn_randomized
 end learn_randomized
 start learn_randomized
 end learn_randomized
 start learn_randomized
 end learn_randomized
 start learn_randomized
 end learn_randomized
 start learn_randomized
 end learn_randomized
 start learn_randomized
 end learn_randomized
 start learn_randomized
 end learn_randomized
 start learn_randomized
 end learn_randomized
 start learn_randomized
 end learn_randomized
 start learn_randomized
 end learn_randomized
 finished cross evals

 Data set contains 387 decoy and 387 target groups.
 mean q_value = 5.898313e-02, std_dev q_value = 1.798605e-02
 mean s_value = 9.797919e-01, std_dev s_value = 1.129259e-01
 calculated scoring and statistics
 processing input data finished
 time needed: 00:00:0.8
==================================================================================================
   qvalue        pvalue    svalue           pep           fdr       fnr           fpr          tp  \
0    0.00  3.683720e-13  0.011107  7.285256e-09  2.475062e-12  0.929828  3.683720e-13    4.000000
1    0.01  1.616470e-01  1.000000  8.532929e-01  1.183753e-02  0.000000  1.616470e-01  362.655626
2    0.02  2.450137e-01  1.000000  1.000000e+00  1.784529e-02  0.000000  2.450137e-01  362.415088
3    0.05  6.980564e-01  1.000000  1.000000e+00  4.924080e-02  0.000000  6.980564e-01  362.239255
4    0.10  9.814054e-01  1.000000  1.000000e+00  6.815489e-02  0.000000  9.814054e-01  360.624058
5    0.20           NaN       NaN           NaN           NaN       NaN           NaN         NaN
6    0.30           NaN       NaN           NaN           NaN       NaN           NaN         NaN
7    0.40           NaN       NaN           NaN           NaN       NaN           NaN         NaN
8    0.50           NaN       NaN           NaN           NaN       NaN           NaN         NaN

          tn            fp          fn    cutoff
0  26.875685  9.900250e-12  356.124315  7.172430
1  22.531311  4.344374e+00   -2.531311  0.987711
2  20.290774  6.584912e+00   -2.290774  0.690265
3   8.114940  1.876074e+01   -2.114940 -0.518819
4   0.499743  2.637594e+01   -0.499743 -2.083680
5        NaN           NaN         NaN       NaN
6        NaN           NaN         NaN       NaN
7        NaN           NaN         NaN       NaN
8        NaN           NaN         NaN       NaN
==================================================================================================
WRITTEN: test_data_ms2_summary_stat.csv
WRITTEN: test_data_ms2_full_stat.csv
WRITTEN: test_data_ms2_scored.tsv
WRITTEN: test_data_ms2_report.pdf
WRITTEN: test_data_ms2_weights.csv

